beacon_node/beacon_chain/src/data_availability_checker.rs

```diff
@@ -511,6 +512,114 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
+    ) -> Result<Option<AvailabilityAndReconstructedColumns<T::EthSpec>>, AvailabilityCheckError>
+    {
+        // Clone the pending components, so we don't hold the read lock during reconstruction
+        let Some(pending_components) = self
+            .availability_cache
+            .peek_pending_components(block_root, |pending_components_opt| {
+                pending_components_opt.cloned()
+            })
+        else {
+            // Block may have been imported as it does not exist in availability cache.
+            return Ok(None);
+        };
+
+        if !self.should_reconstruct(&pending_components) {
+            return Ok(None);
+        }
+
+        self.availability_cache
+            .set_reconstruction_started(block_root);
+
+        metrics::inc_counter(&KZG_DATA_COLUMN_RECONSTRUCTION_ATTEMPTS);
+        let timer = metrics::start_timer(&metrics::DATA_AVAILABILITY_RECONSTRUCTION_TIME);
+
+        let all_data_columns = KzgVerifiedCustodyDataColumn::reconstruct_columns(
+            &self.kzg,
+            &pending_components.verified_data_columns,
+            &self.spec,
+        )
+        .map_err(|e| {
+            error!(
+                self.log,
+                "Error reconstructing data columns";
+                "block_root" => ?block_root,
+                "error" => ?e
+            );
+            metrics::inc_counter(&KZG_DATA_COLUMN_RECONSTRUCTION_FAILURES);
+            AvailabilityCheckError::ReconstructColumnsError(e)
+        })?;
+
+        // Check indices from cache again to make sure we don't publish components we've already received.
+        let Some(existing_column_indices) = self.imported_custody_column_indexes(block_root) else {
+            // If block is already imported (no longer in cache), abort publishing data columns
+            // TODO(das) This assumes only supernodes do reconstructions (i.e. custody
+            // requirement = all columns). This behaviour is likely to change in the future when we
+            // get to 2D PeerDAS.
+            return Ok(None);
+        };
+
+        let data_columns_to_publish = all_data_columns
+            .into_iter()
+            .filter(|d| !existing_column_indices.contains(&d.index()))
+            .collect::<Vec<_>>();
+
+        // Return if there's no data columns to publish.
+        let Some(slot) = data_columns_to_publish
+            .first()
+            .map(|d| d.as_data_column().slot())
+        else {
+            return Ok(None);
+        };
+
+        metrics::stop_timer(timer);
+        metrics::inc_counter_by(
+            &metrics::DATA_AVAILABILITY_RECONSTRUCTED_COLUMNS,
+            data_columns_to_publish.len() as u64,
+        );
+
+        debug!(self.log, "Reconstructed columns";
+            "count" => data_columns_to_publish.len(),
+            "block_root" => ?block_root,
+            "slot" => slot,
+        );
+
+        self.availability_cache
+            .put_kzg_verified_data_columns(
```

Comment:
We have the same issue here, we take a snapshot of an old view of the shared state in `imported_custody_column_indexes` and then later acquire the write lock again in `put_kzg_verified_data_columns`. However, if the bug above is fixed and reconstruction strictly happens once we should be fine? Also worst case we publish the same thing twice. However, now that the entry into the da_checker is not dropped until after import this can still be problematic and lead to double imports?

Reply:
> However, if the bug above is fixed and reconstruction strictly happens once we should be fine

Yep I think so. The double import issue has been raised here, I think it might be best to address it separately:
https://github.com/sigp/lighthouse/issues/6439